File: examples/java/ConnectorLifecycleExample.java
package com.nphase.redcapcloud.examples;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Map;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

/**
 * Example demonstrating the complete connector lifecycle:
 * 1. Create connector (with schedule) - automatically creates Prefect deployment
 * 2. Update connector schedule - updates Prefect deployment schedule
 * 3. Trigger manual run - executes flow immediately
 * 4. Deactivate connector - deletes Prefect deployment
 * 5. Delete connector - removes connector and deployment
 * 
 * Prerequisites:
 * - Backend API running and accessible
 * - Valid authentication token (if required)
 * - Prefect Server running and accessible from backend
 */
public class ConnectorLifecycleExample {
    
    private static final String BASE_URL = "http://localhost:8080/api/v1";
    private static final String AUTH_TOKEN = "your-jwt-token-here"; // Optional if auth is disabled
    private static final HttpClient httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(10))
            .build();
    private static final ObjectMapper objectMapper = new ObjectMapper();
    
    public static void main(String[] args) {
        try {
            System.out.println("=== Connector Lifecycle Example ===\n");
            
            // Step 1: Create Connector (with schedule)
            Long connectorId = createConnector();
            
            if (connectorId == null) {
                System.err.println("Failed to create connector. Exiting.");
                return;
            }
            
            // Wait a moment for deployment to be created
            Thread.sleep(2000);
            
            // Step 2: Update Connector Schedule
            updateConnectorSchedule(connectorId, "0 */6 * * *"); // Change to every 6 hours
            
            // Step 3: Trigger Manual Run
            triggerManualRun(connectorId);
            
            // Step 4: Deactivate Connector (deletes deployment)
            deactivateConnector(connectorId);
            
            // Step 5: Delete Connector
            deleteConnector(connectorId);
            
            System.out.println("\n=== Lifecycle Example Complete ===");
            
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
    
    /**
     * Step 1: Create a new connector with a schedule.
     * This automatically creates a Prefect deployment via the backend.
     */
    private static Long createConnector() throws Exception {
        System.out.println("Step 1: Creating connector with schedule...");
        
        ObjectNode requestBody = objectMapper.createObjectNode();
        requestBody.put("name", "Example Customer Data Pipeline");
        requestBody.put("description", "Example connector for demonstration");
        requestBody.put("projectId", 1L); // Update with your project ID
        requestBody.put("definedSourceId", 1L); // Update with your source ID
        requestBody.put("definedDestinationId", 1L); // Update with your destination ID
        requestBody.put("schedule", "0 */4 * * *"); // Every 4 hours
        requestBody.put("isActive", true);
        requestBody.set("sourceConfigOverride", objectMapper.createObjectNode());
        requestBody.set("destinationConfigOverride", objectMapper.createObjectNode());
        
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/connectors"))
                .header("Content-Type", "application/json")
                .header("Authorization", "Bearer " + AUTH_TOKEN) // Optional
                .POST(HttpRequest.BodyPublishers.ofString(requestBody.toString()))
                .build();
        
        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        
        if (response.statusCode() == 200 || response.statusCode() == 201) {
            ObjectNode responseBody = (ObjectNode) objectMapper.readTree(response.body());
            Long connectorId = responseBody.get("id").asLong();
            String deploymentId = responseBody.has("prefectDeploymentId") 
                    ? responseBody.get("prefectDeploymentId").asText() 
                    : "pending";
            
            System.out.println("✅ Connector created successfully!");
            System.out.println("   Connector ID: " + connectorId);
            System.out.println("   Prefect Deployment ID: " + deploymentId);
            System.out.println("   Schedule: 0 */4 * * * (every 4 hours)");
            System.out.println("   Note: Prefect deployment is created automatically by the backend\n");
            
            return connectorId;
        } else {
            System.err.println("❌ Failed to create connector: " + response.statusCode());
            System.err.println("Response: " + response.body());
            return null;
        }
    }
    
    /**
     * Step 2: Update the connector's schedule.
     * This automatically updates the Prefect deployment schedule via the backend.
     */
    private static void updateConnectorSchedule(Long connectorId, String newSchedule) throws Exception {
        System.out.println("Step 2: Updating connector schedule...");
        
        ObjectNode requestBody = objectMapper.createObjectNode();
        requestBody.put("schedule", newSchedule);
        
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/connectors/" + connectorId))
                .header("Content-Type", "application/json")
                .header("Authorization", "Bearer " + AUTH_TOKEN) // Optional
                .method("PATCH", HttpRequest.BodyPublishers.ofString(requestBody.toString()))
                .build();
        
        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        
        if (response.statusCode() == 200) {
            System.out.println("✅ Connector schedule updated successfully!");
            System.out.println("   New Schedule: " + newSchedule + " (every 6 hours)");
            System.out.println("   Note: Prefect deployment schedule updated automatically by the backend\n");
        } else {
            System.err.println("❌ Failed to update schedule: " + response.statusCode());
            System.err.println("Response: " + response.body());
        }
    }
    
    /**
     * Step 3: Trigger a manual run of the connector.
     * This executes the flow immediately without waiting for the schedule.
     */
    private static void triggerManualRun(Long connectorId) throws Exception {
        System.out.println("Step 3: Triggering manual run...");
        
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/connectors/" + connectorId + "/trigger"))
                .header("Authorization", "Bearer " + AUTH_TOKEN) // Optional
                .POST(HttpRequest.BodyPublishers.noBody())
                .build();
        
        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        
        if (response.statusCode() == 200) {
            System.out.println("✅ Manual run triggered successfully!");
            System.out.println("   Response: " + response.body());
            System.out.println("   Note: Flow is executing in Prefect Worker\n");
        } else {
            System.err.println("❌ Failed to trigger manual run: " + response.statusCode());
            System.err.println("Response: " + response.body());
        }
    }
    
    /**
     * Step 4: Deactivate the connector.
     * This automatically deletes the Prefect deployment via the backend.
     */
    private static void deactivateConnector(Long connectorId) throws Exception {
        System.out.println("Step 4: Deactivating connector...");
        
        ObjectNode requestBody = objectMapper.createObjectNode();
        requestBody.put("isActive", false);
        
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/connectors/" + connectorId))
                .header("Content-Type", "application/json")
                .header("Authorization", "Bearer " + AUTH_TOKEN) // Optional
                .method("PATCH", HttpRequest.BodyPublishers.ofString(requestBody.toString()))
                .build();
        
        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        
        if (response.statusCode() == 200) {
            System.out.println("✅ Connector deactivated successfully!");
            System.out.println("   Note: Prefect deployment deleted automatically by the backend");
            System.out.println("   Note: Scheduled runs will no longer execute\n");
        } else {
            System.err.println("❌ Failed to deactivate connector: " + response.statusCode());
            System.err.println("Response: " + response.body());
        }
    }
    
    /**
     * Step 5: Delete the connector.
     * This automatically deletes the Prefect deployment (if it still exists) via the backend.
     */
    private static void deleteConnector(Long connectorId) throws Exception {
        System.out.println("Step 5: Deleting connector...");
        
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/connectors/" + connectorId))
                .header("Authorization", "Bearer " + AUTH_TOKEN) // Optional
                .DELETE()
                .build();
        
        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        
        if (response.statusCode() == 200 || response.statusCode() == 204) {
            System.out.println("✅ Connector deleted successfully!");
            System.out.println("   Note: Prefect deployment deleted automatically by the backend\n");
        } else {
            System.err.println("❌ Failed to delete connector: " + response.statusCode());
            System.err.println("Response: " + response.body());
        }
    }
}

File: examples/python/connector_lifecycle_example.py
#!/usr/bin/env python3
"""
Example demonstrating the complete connector lifecycle:
1. Create connector (with schedule) - automatically creates Prefect deployment
2. Update connector schedule - updates Prefect deployment schedule
3. Trigger manual run - executes flow immediately
4. Deactivate connector - deletes Prefect deployment
5. Delete connector - removes connector and deployment

Prerequisites:
- Backend API running and accessible
- Valid authentication token (if required)
- Prefect Server running and accessible from backend
- requests library: pip install requests
"""

import requests
import json
import time
from typing import Optional, Dict, Any

# Configuration
BASE_URL = "http://localhost:8080/api/v1"
AUTH_TOKEN = "your-jwt-token-here"  # Optional if auth is disabled
HEADERS = {
    "Content-Type": "application/json",
    "Authorization": f"Bearer {AUTH_TOKEN}"  # Optional
}


def create_connector() -> Optional[int]:
    """
    Step 1: Create a new connector with a schedule.
    This automatically creates a Prefect deployment via the backend.
    
    Returns:
        Connector ID if successful, None otherwise
    """
    print("Step 1: Creating connector with schedule...")
    
    payload = {
        "name": "Example Customer Data Pipeline",
        "description": "Example connector for demonstration",
        "projectId": 1,  # Update with your project ID
        "definedSourceId": 1,  # Update with your source ID
        "definedDestinationId": 1,  # Update with your destination ID
        "schedule": "0 */4 * * *",  # Every 4 hours
        "isActive": True,
        "sourceConfigOverride": {},
        "destinationConfigOverride": {}
    }
    
    try:
        response = requests.post(
            f"{BASE_URL}/connectors",
            headers=HEADERS,
            json=payload,
            timeout=30
        )
        
        if response.status_code in [200, 201]:
            data = response.json()
            connector_id = data.get("id")
            deployment_id = data.get("prefectDeploymentId", "pending")
            
            print("✅ Connector created successfully!")
            print(f"   Connector ID: {connector_id}")
            print(f"   Prefect Deployment ID: {deployment_id}")
            print("   Schedule: 0 */4 * * * (every 4 hours)")
            print("   Note: Prefect deployment is created automatically by the backend\n")
            
            return connector_id
        else:
            print(f"❌ Failed to create connector: {response.status_code}")
            print(f"Response: {response.text}")
            return None
            
    except Exception as e:
        print(f"❌ Error creating connector: {e}")
        return None


def update_connector_schedule(connector_id: int, new_schedule: str) -> bool:
    """
    Step 2: Update the connector's schedule.
    This automatically updates the Prefect deployment schedule via the backend.
    
    Args:
        connector_id: Connector ID
        new_schedule: New cron schedule (e.g., "0 */6 * * *")
    
    Returns:
        True if successful, False otherwise
    """
    print("Step 2: Updating connector schedule...")
    
    payload = {
        "schedule": new_schedule
    }
    
    try:
        response = requests.patch(
            f"{BASE_URL}/connectors/{connector_id}",
            headers=HEADERS,
            json=payload,
            timeout=30
        )
        
        if response.status_code == 200:
            print("✅ Connector schedule updated successfully!")
            print(f"   New Schedule: {new_schedule} (every 6 hours)")
            print("   Note: Prefect deployment schedule updated automatically by the backend\n")
            return True
        else:
            print(f"❌ Failed to update schedule: {response.status_code}")
            print(f"Response: {response.text}")
            return False
            
    except Exception as e:
        print(f"❌ Error updating schedule: {e}")
        return False


def trigger_manual_run(connector_id: int) -> bool:
    """
    Step 3: Trigger a manual run of the connector.
    This executes the flow immediately without waiting for the schedule.
    
    Args:
        connector_id: Connector ID
    
    Returns:
        True if successful, False otherwise
    """
    print("Step 3: Triggering manual run...")
    
    try:
        response = requests.post(
            f"{BASE_URL}/connectors/{connector_id}/trigger",
            headers=HEADERS,
            timeout=60  # Longer timeout for flow execution
        )
        
        if response.status_code == 200:
            print("✅ Manual run triggered successfully!")
            print(f"   Response: {response.text}")
            print("   Note: Flow is executing in Prefect Worker\n")
            return True
        else:
            print(f"❌ Failed to trigger manual run: {response.status_code}")
            print(f"Response: {response.text}")
            return False
            
    except Exception as e:
        print(f"❌ Error triggering manual run: {e}")
        return False


def deactivate_connector(connector_id: int) -> bool:
    """
    Step 4: Deactivate the connector.
    This automatically deletes the Prefect deployment via the backend.
    
    Args:
        connector_id: Connector ID
    
    Returns:
        True if successful, False otherwise
    """
    print("Step 4: Deactivating connector...")
    
    payload = {
        "isActive": False
    }
    
    try:
        response = requests.patch(
            f"{BASE_URL}/connectors/{connector_id}",
            headers=HEADERS,
            json=payload,
            timeout=30
        )
        
        if response.status_code == 200:
            print("✅ Connector deactivated successfully!")
            print("   Note: Prefect deployment deleted automatically by the backend")
            print("   Note: Scheduled runs will no longer execute\n")
            return True
        else:
            print(f"❌ Failed to deactivate connector: {response.status_code}")
            print(f"Response: {response.text}")
            return False
            
    except Exception as e:
        print(f"❌ Error deactivating connector: {e}")
        return False


def delete_connector(connector_id: int) -> bool:
    """
    Step 5: Delete the connector.
    This automatically deletes the Prefect deployment (if it still exists) via the backend.
    
    Args:
        connector_id: Connector ID
    
    Returns:
        True if successful, False otherwise
    """
    print("Step 5: Deleting connector...")
    
    try:
        response = requests.delete(
            f"{BASE_URL}/connectors/{connector_id}",
            headers=HEADERS,
            timeout=30
        )
        
        if response.status_code in [200, 204]:
            print("✅ Connector deleted successfully!")
            print("   Note: Prefect deployment deleted automatically by the backend\n")
            return True
        else:
            print(f"❌ Failed to delete connector: {response.status_code}")
            print(f"Response: {response.text}")
            return False
            
    except Exception as e:
        print(f"❌ Error deleting connector: {e}")
        return False


def get_connector(connector_id: int) -> Optional[Dict[str, Any]]:
    """
    Helper function to retrieve connector details.
    
    Args:
        connector_id: Connector ID
    
    Returns:
        Connector data if successful, None otherwise
    """
    try:
        response = requests.get(
            f"{BASE_URL}/connectors/{connector_id}",
            headers=HEADERS,
            timeout=30
        )
        
        if response.status_code == 200:
            return response.json()
        else:
            print(f"Failed to get connector: {response.status_code}")
            return None
            
    except Exception as e:
        print(f"Error getting connector: {e}")
        return None


def main():
    """Main function demonstrating the complete lifecycle."""
    print("=== Connector Lifecycle Example ===\n")
    
    # Step 1: Create Connector (with schedule)
    connector_id = create_connector()
    
    if connector_id is None:
        print("Failed to create connector. Exiting.")
        return
    
    # Wait a moment for deployment to be created
    time.sleep(2)
    
    # Step 2: Update Connector Schedule
    update_connector_schedule(connector_id, "0 */6 * * *")  # Change to every 6 hours
    
    # Step 3: Trigger Manual Run
    trigger_manual_run(connector_id)
    
    # Step 4: Deactivate Connector (deletes deployment)
    deactivate_connector(connector_id)
    
    # Step 5: Delete Connector
    delete_connector(connector_id)
    
    print("\n=== Lifecycle Example Complete ===")


if __name__ == "__main__":
    main()

